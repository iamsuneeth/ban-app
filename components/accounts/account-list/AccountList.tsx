import React from "react";
import {
  View,
  Text,
  ScrollView,
  Dimensions,
  StyleSheet,
  GestureResponderEvent,
  TouchableHighlight,
  Platform
} from "react-native";
import { TouchableNativeFeedback } from "react-native-gesture-handler";
import { NavigationStackProp } from "react-navigation-stack";
import { Amount } from "../../elements/amount/Amount";
import { Card } from "../../elements/card/Card";

const TestAccounts = [
  {
    section: "Savings",
    accounts: [
      {
        id: "232323",
        accountNumber: "823232323",
        code: "ABCTR050000",
        balance: { amount: 32344.22, currency: "GBP" },
        nickName: "Current Account",
        overdraft: { amount: 323.33, currency: "GBP" },
        availableOverdraft: { amount: 1500, currency: "GBP" },
        availableBalance: { amount: 340, currency: "GBP" },
        openingDate: "2018-09-05",
        customerId: 12567,
        type: "ADVANCE",
        debitCardAvailable: true
      },
      {
        id: "4567",
        accountNumber: "678904455",
        code: "ABCTR050000",
        balance: { amount: 456.5, currency: "GBP" },
        nickName: "Savings Account",
        overdraft: { amount: 45.33, currency: "GBP" },
        availableOverdraft: { amount: 2000, currency: "GBP" },
        availableBalance: { amount: 23, currency: "GBP" },
        openingDate: "2017-09-05",
        customerId: 12567,
        type: "STANDARD",
        debitCardAvailable: false
      }
    ]
  }
];

const getTouchableComponent = ({
  child,
  onPress
}: {
  child: React.ReactNode;
  onPress: (event: GestureResponderEvent) => void;
}) => {
  if (Platform.OS === "android") {
    return (
      <View>
        <TouchableNativeFeedback testID="accountClick" onPress={onPress}>
          {child}
        </TouchableNativeFeedback>
      </View>
    );
  } else {
    return (
      <View>
        <TouchableHighlight testID="accountClick" onPress={onPress}>
          {child}
        </TouchableHighlight>
      </View>
    );
  }
};

type Props = {
  navigation: NavigationStackProp<{}>;
};

export const AccountList = ({ navigation }: Props) => {
  return (
    <View style={styles.accounts}>
      <ScrollView>
        {TestAccounts.map(section => (
          <Card key={section.section}>
            <Text style={styles.sectionHeader}>{section.section}</Text>
            {section.accounts.map((account, index) => {
              return (
                <View key={account.id}>
                  {getTouchableComponent({
                    child: (
                      <View style={styles.accountCard}>
                        <View style={styles.accountPrimary}>
                          <View>
                            <Text style={styles.main}>{account.nickName}</Text>
                            <Text style={styles.secondary}>
                              {account.code + " " + account.accountNumber}
                            </Text>
                          </View>
                          <View>
                            <Amount
                              amount={account.balance.amount}
                              currency={account.balance.currency}
                              size={25}
                            />
                          </View>
                        </View>
                      </View>
                    ),
                    onPress: () =>
                      navigation.navigate("AccountDetails", {
                        accountId: account.id
                      })
                  })}
                  {index !== section.accounts.length - 1 && (
                    <View
                      style={{
                        backgroundColor: "#f5f5f5",
                        height: 1,
                        marginHorizontal: 10
                      }}
                    />
                  )}
                </View>
              );
            })}
          </Card>
        ))}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  accounts: {
    flex: 4,
    width: Dimensions.get("window").width
  },
  accountCard: {
    marginVertical: 5,
    minHeight: 80,
    backgroundColor: "#fff"
  },
  accountPrimary: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between"
  },
  main: { color: "#333", fontSize: 16 },
  secondary: { color: "gray", fontSize: 12 },
  highlight: { color: "#333", fontSize: 18 },
  sectionHeader: {
    fontSize: 20,
    marginVertical: 10,
    paddingBottom: 5,
    fontWeight: "bold",
    color: "gray"
  }
});
