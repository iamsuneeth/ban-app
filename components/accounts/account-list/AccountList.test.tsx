import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { AccountList } from "./AccountList";

jest.mock("react-native-gesture-handler", () => {});

describe("tests for AccountList component", () => {
  const navigation = {
    navigate: jest.fn(element => null)
  };
  test("should render correctly", () => {
    const { baseElement } = render(
      <AccountList navigation={navigation as any} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  test("should call details navigation on account click", () => {
    const { queryAllByTestId } = render(
      <AccountList navigation={navigation as any} />
    );

    fireEvent.press(queryAllByTestId("accountClick")[0]);
    expect(navigation.navigate).toHaveBeenCalledWith("AccountDetails");
  });
});
