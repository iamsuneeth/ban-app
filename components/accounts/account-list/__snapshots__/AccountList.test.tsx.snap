// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for AccountList component should render correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 4,
          "width": 750,
        }
      }
    >
      <ScrollView>
        <View>
          <Text
            style={
              Object {
                "color": "#555",
                "fontSize": 40,
                "fontWeight": "bold",
                "margin": 15,
              }
            }
          >
            Savings
          </Text>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 5,
                  "elevation": 2,
                  "marginHorizontal": 10,
                  "marginVertical": 5,
                  "padding": 10,
                  "shadowColor": "#ccc",
                  "shadowOffset": Object {
                    "height": 3,
                    "width": 0,
                  },
                  "shadowOpacity": 1,
                  "shadowRadius": 5,
                },
                Object {
                  "backgroundColor": "#039be5",
                  "marginVertical": 5,
                  "minHeight": 197,
                  "padding": 0,
                },
              ]
            }
          >
            <Button
              style={
                Object {
                  "flex": 1,
                  "padding": 10,
                }
              }
              testID="accountClick"
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "space-between",
                  }
                }
              >
                <View>
                  <Text
                    style={
                      Object {
                        "color": "#fff",
                        "fontSize": 32,
                      }
                    }
                  >
                    Current Account
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#fff",
                        "fontSize": 24,
                      }
                    }
                  >
                    ABCTR050000 823232323
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "alignSelf": "flex-end",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-end",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "#fff",
                          "fontSize": 50,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      £32,344
                    </Text>
                    <Text
                      style={
                        Object {
                          "color": "#fff",
                          "fontSize": 40,
                          "paddingBottom": 1,
                        }
                      }
                      testID="decimal"
                    >
                      .
                      22
                    </Text>
                  </View>
                </View>
              </View>
            </Button>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 5,
                  "elevation": 2,
                  "marginHorizontal": 10,
                  "marginVertical": 5,
                  "padding": 10,
                  "shadowColor": "#ccc",
                  "shadowOffset": Object {
                    "height": 3,
                    "width": 0,
                  },
                  "shadowOpacity": 1,
                  "shadowRadius": 5,
                },
                Object {
                  "backgroundColor": "#039be5",
                  "marginVertical": 5,
                  "minHeight": 197,
                  "padding": 0,
                },
              ]
            }
          >
            <Button
              style={
                Object {
                  "flex": 1,
                  "padding": 10,
                }
              }
              testID="accountClick"
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "space-between",
                  }
                }
              >
                <View>
                  <Text
                    style={
                      Object {
                        "color": "#fff",
                        "fontSize": 32,
                      }
                    }
                  >
                    Savings Account
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#fff",
                        "fontSize": 24,
                      }
                    }
                  >
                    ABCTR050000 678904455
                  </Text>
                </View>
                <View
                  style={
                    Object {
                      "alignSelf": "flex-end",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-end",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "color": "#fff",
                          "fontSize": 50,
                          "fontWeight": "bold",
                        }
                      }
                    >
                      £456
                    </Text>
                    <Text
                      style={
                        Object {
                          "color": "#fff",
                          "fontSize": 40,
                          "paddingBottom": 1,
                        }
                      }
                      testID="decimal"
                    >
                      .
                      22
                    </Text>
                  </View>
                </View>
              </View>
            </Button>
          </View>
        </View>
      </ScrollView>
    </View>
  </View>
</View>
`;
